### DISCUSSION TITLE
[Feature] standardize the `dbt.type_timestamp` macro

### DISCUSSION THREAD

**Discussion by fivetran-reneeli** [dbt-labs/dbt-core#9790] [Ideas]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Not sure if this is a bug, but I am wondering if there are plans to standardize the [`dbt.type_timestamp`](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros#type_timestamp) macro. I believe in most warehouses, it defaults to timestamp without timezone, so it would be helpful to confirm that this is the end result across all adapters.

### Expected Behavior

Field is a timestamp without time zone

### Steps To Reproduce

n/a

### Relevant log output

```shell
na
```

### Environment

```markdown
- OS: mac
- Python: 3.12.2
- dbt: 1.7.9
```

### Which database adapter are you using with dbt?

other (mention it in "Additional Context")

### Additional Context

Using multiple adapters

**Comment by dbeatty10** [comment]
Thanks for reaching out about this @fivetran-reneeli 🧠

What are the use-case(s) that you are trying to solve for?

Is it doing casts like the following or something else?

```
cast({{ expression }} as {{ dbt.type_timestamp() }})
```

### Short story

Our docs state that [`type_timestamp`](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros#type_timestamp):
> may or may not match the behavior of TIMESTAMP WITHOUT TIMEZONE from ANSI SQL-92

So even though it is inconvenient that it is not standardized across adapters, this is known and documented behavior rather than a bug.

> I am wondering if there are plans to standardize the [`dbt.type_timestamp`](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros#type_timestamp) macro

We don't currently have plans to standardize this.

### Longer story

Here are some other issues/comments that may be related:
- https://github.com/calogica/dbt-expectations/issues/192
- https://github.com/dbt-labs/dbt-core/issues/6453
- https://github.com/dbt-labs/metricflow/issues/733#issuecomment-1726411113

#### Possibility: a new `type_timestamp_ntz()` macro

If we were to standardize this, one option would be to introduce a new macro in order to maintain backwards compatibility.

Let's suppose there's a new macro called `type_timestamp_ntz`. Here's the strings it might return on a per adapter basis to align with naive timestamps that align with `TIMESTAMP WITHOUT TIME ZONE` from the SQL standard (or `java.time.LocalDateTime` from Java 8):

| Database       | Naive timestamp data type               |
|-----------------|-----------------------------|
| Starburst/Trino | `TIMESTAMP WITHOUT TIME ZONE` |
| Athena          | `TIMESTAMP WITHOUT TIME ZONE` |
| Postgres        | `TIMESTAMP`                   |
| Redshift        | `TIMESTAMP`                   |
| AlloyDB         | `TIMESTAMP`                   |
| Dremio          | `TIMESTAMP`                   |
| Oracle          | `TIMESTAMP`                   |
| Snowflake       | `TIMESTAMP_NTZ `              |
| Spark SQL       | `TIMESTAMP_NTZ`               |
| Databricks      | `TIMESTAMP_NTZ`               |
| BigQuery        | `DATETIME`                    |
| Teradata        | N/A                         |

#### Caveats

- This is based off reading the documentation for each database; I didn't actually try them out to double-check each.
- `TIMESTAMP_NTZ` is not available until Spark 3.4 & Databricks Runtime 13 [[1](https://www.linkedin.com/posts/smilegator_introducing-timestampntz-in-apache-spark-activity-7132600585253531648-V4LI/), [2](https://community.databricks.com/t5/technical-blog/introducing-timestamp-ntz-in-apache-spark-the-timestamp-without/ba-p/50586)]
- Teradata has separate `TIMESTAMP` and `TIMESTAMP WITH TIME ZONE` data types, but the former is described as "non-ANSI standard. Analytics Database stores a TIMESTAMP value in UTC." [[3](https://docs.teradata.com/r/Enterprise_IntelliFlex_VMware/SQL-Data-Types-and-Literals/DateTime-and-Interval-Data-Types/TIMESTAMP-Data-Type)]

**Comment by fivetran-reneeli** [comment]
Thanks for the thorough response @dbeatty10 ! It was helpful to have the related issues-- I found [this one](https://github.com/dbt-labs/dbt-core/issues/6453) to be very in line with extra attention on your comment [here](https://github.com/dbt-labs/dbt-core/issues/6453#issuecomment-1445431118). 

> It would restore the original literal string values returned bytype_timestamp() that [existed in dbt-utils](https://github.com/dbt-labs/dbt-utils/blob/ac072a3c4b78d43a1c013e7de8b8fa6e290b544e/macros/cross_db_utils/datatypes.sql#L25-L41) previously (e.g., TIMESTAMP_NTZ instead of TIMESTAMP for Snowflake, etc)

For more context, on our side our models leverage dbt.type_timestamp for the following warehouses:
Snowflake, Postgres, Databricks, Redshift, and BigQuery.

So like you listed, I believe each one has default TNZ except for BigQuery, where `timestamp` = with timezone, and `datetime` = without timezone. 

The dbt team may have discussed this in the past, based on comments I found like [this](https://github.com/dbt-labs/dbt-core/blob/8a395e928d1016368712e80641642a57e59590b4/tests/functional/adapter/utils/data_types/test_type_timestamp.py#L8-L9). And I dug around the old dbt_utils [changelog](https://github.com/dbt-labs/dbt-utils/blob/23da1f424398696e25656dc6f89066b9a5d13794/CHANGELOG.md?plain=1#L276) and saw that the timestamp macro for redshift and postgres were explicitly cast once upon a time. 

cc @fivetran-joemarkiewicz

**Comment by dbeatty10** [comment]
@fivetran-reneeli All those links you included are golden info 🤩

> For more context, on our side our models leverage dbt.type_timestamp for the following warehouses:
> Snowflake, Postgres, Databricks, Redshift, and BigQuery.

Could you share more about how you are leveraging `dbt.type_timestamp`? e.g. could you share the end goal that you are trying to accomplish by using it? If you have links to source code, that would be handy.

The reason I ask is there are at least a few different uses I can think of:
- comparing the data type from an information schema query (less common)
- casting an expression to a naive timestamp (more common)
- casting an expression to an aware timestamp (less common?)

**Comment by fivetran-reneeli** [comment]
Sorry for the delay!

Sure thing, so our team develops data model packages that we make compatible with dbt, that are warehouse-agnostic (so currently the packages work across postgres, snowflake, bigquery, redshift, and databricks). [Here](https://github.com/fivetran/dbt_jira_source/blob/e78848d2dd476e54bc4744356c6c7d36dff56efa/models/stg_jira__issue.sql#L27) is an example model with jira.

`cast(created as {{ dbt.type_timestamp() }}) as created_at,`

Because we want our models to work with all those mentioned warehouses, ideally the compiled code is the same for all.

**Comment by dbeatty10** [comment]
No prob and thanks for that example @fivetran-reneeli !

There's some tricky things to consider here. For example, how would this example behave when `created` expression in is an aware timestamp vs. a naive timestamp?

We took a look at something similar a year or so ago when considering https://github.com/dbt-labs/dbt-core/issues/5969, and we opted not to tackle it at that time due to the complexity of getting it right and the high probability of accidentally getting it wrong and creating churn for users and maintainers.

See https://github.com/dbt-labs/dbt-core/pull/5935 and https://github.com/dbt-labs/dbt-core/pull/7665 for a small sample of edge cases related to aware vs. naive that we've run into.

So I'm going to upgrade this issue to a Discussion so we can further discuss different design aspirations folks are looking for as well as edge cases people have come across.

I'd propose the main topic of the discussion to be this question:
- How can we best support converting timestamp expressions to naive (and/or aware) timestamps in a cross-database manner?
